aten/src/ATen/cuda/CUDAContext.cpp

```diff
@@ -1,19 +1,48 @@
 #include "ATen/cuda/CUDAContext.h"
 #include "THC/THCGeneral.hpp"
 
+#include <mutex>
+#include <deque>
+#include <vector>
+
 namespace at { namespace cuda {
 
+namespace {
+
+static DeviceIndex num_gpus = -1;
+static std::once_flag init_flag;
+static std::deque<std::once_flag> device_flags;
```

Comment:
I guess, you have a good reason for not wanting to eagerly read out device properties from each device?

Reply:
What do you mean by eagerly in this case? Just making the cudaGetDeviceProperties call whenever it's needed?

I was following the std::call_once pattern from CUDAStream.cpp to achieve eager initialization. Also, since THCudaInit cached deviceProperties for all GPUs, I wanted to keep the same functionality by caching deviceProperties for each GPUs in a std::vector. I don't know how much overhead cudaGetDeviceProperties has every time we make a call to it, and hence, just wanted to keep it cached.